File: src/org/irislang/jiris/dev/IrisClassRoot.java
package org.irislang.jiris.dev;

import java.lang.invoke.CallSite;
import java.lang.invoke.ConstantCallSite;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;

import org.irislang.jiris.core.IrisClass;
import org.irislang.jiris.core.IrisModule;
import org.irislang.jiris.core.exceptions.IrisExceptionBase;

public abstract class IrisClassRoot {

	public static CallSite BootstrapMethod(Class<?> classObj, MethodHandles.Lookup lookup, String name, MethodType mt) throws Throwable {
        return new ConstantCallSite(lookup.findStatic(classObj, name, mt));
    }
	
	abstract public String NativeClassNameDefine();
	abstract public IrisClass NativeSuperClassDefine();
	abstract public IrisModule NativeUpperModuleDefine();
	abstract public Object NativeAlloc();
	abstract public void NativeClassDefine(IrisClass classObj) throws IrisExceptionBase;
}

File: src/main/java/org/irislang/jiris/dev/IrisClassRoot.kt
package org.irislang.jiris.dev

import java.lang.invoke.CallSite
import java.lang.invoke.ConstantCallSite
import java.lang.invoke.MethodHandles
import java.lang.invoke.MethodType

import org.irislang.jiris.core.IrisClass
import org.irislang.jiris.core.IrisModule
import org.irislang.jiris.core.exceptions.IrisExceptionBase

abstract class IrisClassRoot {

    abstract fun NativeClassNameDefine(): String
    abstract fun NativeSuperClassDefine(): IrisClass
    abstract fun NativeUpperModuleDefine(): IrisModule?
    abstract fun NativeAlloc(): Any
    @Throws(IrisExceptionBase::class)
    abstract fun NativeClassDefine(classObj: IrisClass): Any

    companion object {

        @Throws(Throwable::class)
        @JvmStatic fun BootstrapMethod(classObj: Class<*>, lookup: MethodHandles.Lookup, name: String, mt: MethodType): CallSite {
            return ConstantCallSite(lookup.findStatic(classObj, name, mt))
        }
    }
}